File: framework/ruby/src/soda/unittest/feature-factory.rb
require_relative 'feature'
require_relative 'unordered'

module Soda end

module Soda::Unittest

class ListFeature < ObjectFeature
  def initialize(elem_feat)
    @elem_feat = elem_feat
  end

  def get_hash(obj)
    res = 0
    # keep low 48 bits
    mask = 0xffffffffffff
    obj.each { |e|
      h = @elem_feat.get_hash(e)
      res = res * 133 + h
      res &= mask
    }
    res
  end

  def is_equal?(a, b)
    if a.size != b.size
      return false
    end
    (0...a.size).each { |i|
      unless @elem_feat.is_equal?(a[i], b[i])
        return false
      end
    }
  end
end

class FloatFeature < ObjectFeature
  def get_hash(val)
    val.hash
  end

  def is_equal?(a, b)
    (a-b).abs < 1e-6
  end
end

class FeatureFactory
  @factory_map = {}
  @factory_map.default = ->() { ObjectFeature.new }
  @factory_map['Float'] = ->() { FloatFeature.new }
  @factory_map['Float[]'] = ->() { ListFeature.new(self.create('Float')) }
  @factory_map['Float[][]'] = ->() { ListFeature.new(self.create('Float[]')) }

  def self.create(obj_type)
    @factory_map[obj_type].call
  end
end

end # module Soda::Unittest

File: framework/python/src/soda/unittest/featurefactory.py
from collections import defaultdict
from typing import *

from .feature import ObjectFeature
from .unordered import UnorderedListFeature

class ListFeature(ObjectFeature):

    def __init__(self, elemFeat: ObjectFeature):
        self.elemFeat = elemFeat

    def hash(self, obj: List[Any]) -> int:
        res = 0
        # keep low 48 bits
        mask = 0xffffffffffff
        for e in obj:
            h = self.elemFeat.hash(e)
            res = res * 133 + h
            res &= mask
        return res

    def isEqual(self, a: List[Any], b: List[Any]) -> bool:
        if len(a) != len(b):
            return False
        for i in range(len(a)):
            if not self.elemFeat.isEqual(a[i], b[i]):
                return False
        return True

class FloatFeature(ObjectFeature):

    def hash(self, val: float) -> int:
        return hash(val)

    def isEqual(self, a: float, b: float) -> bool:
        return abs(a - b) < 1e-6

class FeatureFactory:

    factoryMap = defaultdict(lambda: ObjectFeature)

    @classmethod
    def create(cls, objType: type) -> ObjectFeature:
        return cls.factoryMap[objType]()

    @classmethod
    def register(cls):
        cls.factoryMap[float] = FloatFeature
        cls.factoryMap[List[float]] = lambda : ListFeature(cls.create(float))
        cls.factoryMap[List[List[float]]] = lambda : ListFeature(cls.create(List[float]))

FeatureFactory.register()
